### PR TITLE
Adding config option to add additional SANs to the master's certificate.

### PR DESCRIPTION
**Pull request by hyperbolic2346**
Regenerate certificate if data on certificate changes. This includes IP address and SANs.
Restart API server after updating certificate.

**What this PR does / why we need it**:
This allows users to add addition SAN entries to the certificate generated.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

https://github.com/juju-solutions/bundle-canonical-kubernetes/issues/426
**Special notes for your reviewer**:

**Release note**:

```release-note
Added support for SAN entries in the master node certificate via juju kubernetes-master config.
```

### CODE REVIEW

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py [hunk 1]
```diff
@@ -457,12 +457,38 @@ def send_data(tls):
         'kubernetes.default.svc',
         'kubernetes.default.svc.{0}'.format(domain)
     ]
+
+    # maybe they have extra names they want as SANs
+    extra_sans = hookenv.config('extra_sans')
+    if extra_sans and not extra_sans == "":
+      sans.extend(extra_sans.split())
```
Comment (Cynerva) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: Can you fix the indentation here so it passes flake8? Some of our tests get fussy if flake8 doesn't pass.

```
$ flake8 reactive
reactive/kubernetes_master.py:464:7: E111 indentation is not a multiple of four
```

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py [hunk 2]
```diff
@@ -457,12 +457,38 @@ def send_data(tls):
...
     tls.request_server_cert(common_name, sans, certificate_name)
 
 
+@when('config.changed', 'certificates.available')
+def update_certificate(tls):
+    # I using the config.changed flag instead of something more
+    # specific to try and catch ip changes. Being a little
+    # spammy here is ok because the cert layer checks for
+    # changes to the cert before issuing a new one
+    send_data(tls)
```
Comment (Cynerva) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: I think we can omit this handler. `send_data` is already invoked on every update-status hook (every ~5 min):

```
$ juju ssh kubernetes-master/0 sudo cat /var/log/juju/unit-kubernetes-master-0.log | grep send_data | tail -n 5
2017-10-19 21:12:58 INFO juju-log Invoking reactive handler: reactive/kubernetes_master.py:437:send_data
2017-10-19 21:17:19 INFO juju-log Invoking reactive handler: reactive/kubernetes_master.py:437:send_data
2017-10-19 21:22:42 INFO juju-log Invoking reactive handler: reactive/kubernetes_master.py:437:send_data
2017-10-19 21:27:24 INFO juju-log Invoking reactive handler: reactive/kubernetes_master.py:437:send_data
2017-10-19 21:32:10 INFO juju-log Invoking reactive handler: reactive/kubernetes_master.py:437:send_data
```

`config.changed` wouldn't cover the ip change case, either. It's only set if the application config changes (e.g. `juju config kubernetes-master ...`). I don't think there is a state that captures an IP change, but considering that `send_data` runs frequently already, we're probably good.

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py (hunk 1 above)
Comment (hyperbolic2346) [reply cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: Of course, sorry about that one.

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py (hunk 2 above)
Comment (hyperbolic2346) [reply cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: My understanding from @johnsca was that an ip change would trigger config.changed. If that isn't the case then this is just hitting for no good reason when another config option is changed and I agree it should be removed. I worry that a possible 5 minute lag before you can hit your master isn't good though. At a minimum, I would like to change that to config.changed.extra_sans so we kick it off when you change your SAN config.

Comment (tvansteenburgh) [reply cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: I'm +1 on changing this to config.changed.extra_sans. I don't think we can rely on juju to tell us when the IP has changed, but you could always add the IP to extra_sans manually if you didn't want to wait the 5 minutes for the next update-status hook.

**Comment by Cynerva** [comment]
/approve no-issue
/lgtm

Looks good, thanks @hyperbolic2346

**Pull request merged**